File: asst6/2.java
import java.util.Random;
class Father extends Thread
{
	Random rand = new Random();
	public void run()
	{
		while(true)
		{
			deposit();
		}
	}
	public synchronized void deposit()
	{
		while(Account.balance <= 2000)
		{
			int r = rand.nextInt(200);
			Account.balance += r;
			System.out.println("Father deposited Rs " + r + ". New Balance is Rs " + Account.balance + ".");
		}
	}
}
class Son extends Thread
{
	Random rand = new Random();
	public void run()
	{
		while(true)
		{
			withdraw();	
		}
	}
	public synchronized void withdraw()
	{
		
		while(Account.balance >= 500)
		{
			int r = rand.nextInt(150);
			Account.balance -= r;
			System.out.println("Son withdrawed Rs " + r + ". New Balance is Rs " + Account.balance + ".");
		}
	}
}
class Bank
{
	public static void main(String[] args)
	{
		//public int balance = 600;
		Father f = new Father();
		Son s = new Son();
		f.start();
		s.start();
	}
}

File: asst6/2.cpp
#include<bits/stdc++.h>
#include<pthread.h>
#include<mutex>
#include <unistd.h>
#include <stdlib.h>
//g++ -std=c++11 2.cpp -lpthread

using namespace std;
int balance=600;
mutex m;
void * father(void *args)
{
	while(1)
	{
		m.lock();
		while(balance<=2000)
		{
			int x=rand()%200+1;
			balance+=x;
			cout<<"Amount deposited by father :"<<x<<"\n";
			cout<<"New balance :"<<balance<<"\n";
		}
		m.unlock();
		sleep(1);
	}
		
}
void *son(void *args)
{

	while(1)
	{
		m.lock();
		while(balance>=500)
		{
			int x=rand()%150+1;
			balance-=x;
			cout<<"Amount withdraw by son :"<<x<<"\n";
			cout<<"New balance :"<<balance<<"\n";
		}
		m.unlock();
		sleep(1);	
	}
}

int main()
{
	pthread_t f,s;
	pthread_create(&f,NULL,&father,NULL);
	pthread_create(&s,NULL,&son,NULL);

	pthread_join(f,NULL);
	pthread_join(s,NULL);
	return 0;
}
